src/parser/expr.rs

```diff
@@ -1,18 +1,136 @@
+use std::cmp::Ordering;
+
+use crate::ice::Ice;
 use crate::lexer::token::Token;
 use crate::parser::ast::{Expr, ExprKind, Ident, Lit};
 use crate::parser::error::{ParseError, ParseResult};
 use crate::parser::span::{Spanned, WithSpan};
 use crate::parser::Parser;
 use bumpalo::collections::Vec;
 
-use super::ast::{BinOp, UnaryOp};
+use super::ast::{BinaryOp, UnaryOp};
 use super::span::Span;
 
+impl ToString for BinaryOp {
+    fn to_string(&self) -> String {
+        match self {
+            BinaryOp::Mul => "+",
+            BinaryOp::Div => "/",
+            BinaryOp::Add => "+",
+            BinaryOp::Sub => "-",
+            BinaryOp::LogicalAnd => "&&",
+            BinaryOp::LogicalOr => "||",
+            BinaryOp::Gt => ">",
+            BinaryOp::Gte => ">=",
+            BinaryOp::Lt => "<",
+            BinaryOp::Lte => "<=",
+            BinaryOp::Eq => "==",
+            BinaryOp::Neq => "!=",
+        }
+        .to_string()
+    }
+}
+
+/// Precedence of binary operators
+///
+/// The order of the variants of this enum is significant, because it defines
+/// derived implementation of `PartialOrd`/`Ord`.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub enum Precedence {
+    Lowest,
+
+    /// The precedence of logical disjunction.
+    Disjunction,
+
+    /// The precedence of logical conjunction.
+    Conjunction,
+
+    /// The precedence of comparison operators.
+    Comparison,
+
+    /// The precedence of addition and subtraction.
+    AddSub,
+
+    /// The precedence of multiplication and division.
+    MulDiv,
+}
+
+/// Associativity of binary operator.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Associativity {
+    /// Left associativity.
+    ///
+    /// `<x> o <y> o <z>` is parsed as `(<x> o <y>) o <z>`.
+    Left,
+
+    /// Right associativity.
+    ///
+    /// `<x> o <y> o <z>` is parsed as `<x> o (<y> o <z>)`.
+    #[allow(unused)]
+    Right,
+
+    /// Incompatible operators
+    Not,
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub enum Compatibility {
```

Comment:
`Compatibility` carries exactly the same information as `Associativity` - just renaming its variants.  I think it'd be better to remove `Compatibility` and only use `Associativity`, perhaps adding some more documentation to justify that usage.

Reply:
No, because it has different meaning

Reply:
But yes docs are a good idea